### PR TITLE
improve meta tags:

### PR DESCRIPTION
**Pull request by salim-b**
- fix `author` and `og:url` meta tags
- add missing (Facebook) Open Graph meta tags
    - add `og:locale`
    - add `og:updated_time`
    - make `og:type` an `article` for blog posts (instead of `website`)
    - add `article:` meta tags for blog posts (use the first category of blog
posts for `article:section`)
    - add `og:image:type` MIME type
    - add auto-calculated `og:image:width` and `og:image:height`
    - you can now set the key `facebook_author = "xyz"` in the front matter of
posts (also possible to set a fallback under `Site.Params.facebook_author`)
to have the `article:author` meta tag set to
`"https://www.facebook.com/xyz/"`
    - add `article:publisher` if `facebook_site` is set in `Site.Params`
- use the post banner image for both `og:image` and `twitter:image` if
present and fall back to `static/img/sharing-default.png` otherwise
- make `twitter:card` only of type `summary_large_image` if page is a blog
post and has a dedicated banner image set, otherwise use smaller type `summary`
- add `twitter:creator` meta tag if the key `twitter_author` is present in
the front matter of posts (also possible to set a fallback under
`Site.Params.twitter_author`)
- added some examples for facebook/twitter authors in the example `.md`
posts
- make `twitter:site` meta tag disappear if `twitter_site` is not set in
`Site.Params`
- truncate `twitter:title` and `twitter:description` to max allowed length
(70 and 200 chars)
- `markdownify` and `plainify` various strings for convenience
- simplify code/improve style

### CODE REVIEW

**Comment by highergroundstudio** [comment]
I moved over to using the hugo [internal templates ](https://gohugo.io/templates/internal/) instead as it makes head.html much simpler and outputs the same tags as the FB og and Twitter cards. This is all that you need in your head.html:

```
  <!-- Twitter Card meta tags -->
  {{ template "_internal/twitter_cards.html" . }}

  <!-- Facebook OpenGraph tags -->
  {{ template "_internal/opengraph.html" . }}
```

**Comment by salim-b** [comment]
> I moved over to using the hugo internal templates instead as it makes head.html much simpler and outputs the same tags as the FB og and Twitter cards.

Wow, I didn't even know about those internal templates... thanks for the hint.

Since you already tested Open Graph template, does it automatically set `og:type` to `article` instead of `website` and add all the relevant `article:` meta tags if the page is a blog post?

**Comment by highergroundstudio** [comment]
> > I moved over to using the hugo internal templates instead as it makes head.html much simpler and outputs the same tags as the FB og and Twitter cards.
> 
> Wow, I didn't even know about those internal templates... thanks for the hint!
> 
> Since you already tested the Open Graph template, does it automatically set `og:type` to `article` instead of `website` and add all the relevant `article:` meta tags if the page is a blog post?

Yes, it does both of those.

**Comment by ryanfox1985** [comment]
@salim-b this PR it's very big with different stuff, can you split this PR in small PRs? and focus in their topics.

**Comment by salim-b** [comment]
> @salim-b this PR is very big with different stuff, can you split this PR in small PRs? and focus in their topics.

1. So you want to continue maintaining your own Open Graph / Twitter meta tag implementation? As @highergroundstudio [pointed out](https://github.com/devcows/hugo-universal-theme/pull/212#issuecomment-486330168), it would be much simpler to switch over to using Hugo's [internal templates](https://gohugo.io/templates/internal/) instead. (I don't have any experience in using them but maybe @highergroundstudio we be kind enough to directly submit a suitable PR since he's already using them)

2. Supposed you don't want to switch to Hugo's internal templates and we hold on to to improving our own meta tag implementation: I honestly think it doesn't make sense to split this PR up into smaller ones because _all_ the changes are related to OG/Twitter meta tags and the code relies on "shared" variables (`$description`, `$is_blog`, `$has_image`, `$image`) at several points. I rather think it would be more complicated to understand what the code does and which meta tags are connected in what ways if I would split it up.

File: layouts/partials/head.html [hunk 1]
```diff
@@ -3,40 +3,31 @@
...
+  {{ $description := default .Site.Params.defaultDescription .Description }}
+  <meta name="description" content="{{ $description | markdownify | plainify }}">
 
   {{ .Hugo.Generator }}
 
+  <!-- Font -->
   <link href='//fonts.googleapis.com/css?family=Roboto:400,100,100italic,300,300italic,500,700,800' rel='stylesheet' type='text/css'>
 
-  <!-- Bootstrap and Font Awesome css -->
+  <!-- Bootstrap and Font Awesome CSS -->
```
Comment (ryanfox1985) [inline layouts/partials/head.html]: I agree with the headers additions or comments refactors, but it's not related with this PR.

File: layouts/partials/head.html [hunk 2]
```diff
@@ -3,40 +3,31 @@
...
-  {{ else if .Site.Params.DefaultKeywords }}
-  <meta name="keywords" content="{{ delimit .Site.Params.DefaultKeywords ", " }}">	
-  {{ end }}
-
-  {{ if .Description }}
-  <meta name="description" content="{{ .Description }}">
-  {{ else if .Site.Params.DefaultDescription }}
-  <meta name="description" content="{{ .Site.Params.DefaultDescription }}">	
-  {{ end }}
+  <title>{{ .Title | markdownify | plainify }}</title>
```
Comment (ryanfox1985) [inline layouts/partials/head.html]: these meta changes are not related with the PR

File: layouts/partials/head.html [hunk 3]
```diff
@@ -48,28 +39,54 @@
...
-  <meta property="og:type" content="website" />
-  <meta property="og:url" content="{{ .Permalink }}/" />
-  <meta property="og:image" content="{{ .Site.Params.logo }}" />
+  {{ $is_blog := and (eq .Type "blog") (eq .Kind "page") }}
+  {{ $has_image :=  isset .Params "banner" }}
+  {{ $image := cond $has_image .Params.banner .Site.Params.default_sharing_image }}
+  {{ $image_ext := path.Ext $image }}
+  <meta property="og:locale" content="{{ replace .Site.LanguageCode "-" "_" }}">
+  <meta property="og:site_name" content="{{ .Site.Title }}">
+  <meta property="og:title" content="{{ .Title | markdownify | plainify }}">
```
Comment (ryanfox1985) [inline layouts/partials/head.html]: repeated code {{ .Title | markdownify | plainify }}

Comment (ryanfox1985) [reply layouts/partials/head.html]: same other like description

File: layouts/partials/head.html (hunk 2 above)
Comment (salim-b) [reply layouts/partials/head.html]: ok, you're right, they're not related to OG/Twitter tags. But honestly: It's just a bunch of senseless additional work to split that up if you're willing to integrate those additional changes anyway. And they're _really not that complicated_ to understand.

So instead: Please just tell me which changes you don't approve and I can remove them.

File: layouts/partials/head.html [hunk 4]
```diff
@@ -3,40 +3,31 @@
...
   <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
 
-  <!-- Css animations  -->
+  <!-- CSS animations  -->
   <link href="{{ "css/animate.css" | relURL}}" rel="stylesheet">
 
   <!-- Theme stylesheet, if possible do not edit this stylesheet -->
-  {{ if and (isset .Site.Params "style") .Site.Params.style }}
-    <link href="{{ "css/style" | relURL}}.{{ .Site.Params.style }}.css" rel="stylesheet" id="theme-stylesheet">
+  {{ with .Site.Params.style }}
```
Comment (ryanfox1985) [inline layouts/partials/head.html]: refactor not related with this PR

Comment (salim-b) [reply layouts/partials/head.html]: I've mentioned it in the PR as "simplify code" :wink:

File: layouts/partials/head.html (hunk 1 above)
Comment (salim-b) [reply layouts/partials/head.html]: You're starting troll here...

File: exampleSite/config.toml [hunk 5]
```diff
@@ -70,9 +70,10 @@ paginate = 10
     defaultKeywords = ["devows", "hugo", "go"]
     defaultDescription = "Site template made by devcows using hugo"
 
-    # Twitter
-    twitter = "GoHugoIO"
-    twitterImage = "img/twitter-default.png"
+    # Social media
```
Comment (ryanfox1985) [inline exampleSite/config.toml]: the readme needs to be modified according these new parameter, also with [twitter/twitter]_author.

File: layouts/partials/head.html (hunk 3 above)
Comment (salim-b) [reply layouts/partials/head.html]: ok, agree.

File: layouts/partials/head.html [hunk 6]
```diff
@@ -48,28 +39,54 @@
...
+  <meta property="og:description" content="{{ $description | markdownify | plainify }}">
+  <meta property="og:image" content="{{ $image | absURL }}">
+  <meta property="og:image:type" content="image/{{ if eq $image_ext ".svg" }}svg+xml{{ else }}{{ trim $image_ext "." }}{{ end }}">
+  {{ with .Params.banner_alt }}<meta property="og:image:alt" content="{{ . | markdownify | plainify }}">{{ end }}
+  {{ $image_local :=  printf "/static/%s" $image}}
+  {{ with (imageConfig $image_local) }}
+    <meta property="og:image:width" content="{{ .Width }}">
+    <meta property="og:image:height" content="{{ .Height }}">
+  {{ end }}
+  {{ with .Lastmod }}<meta property="og:updated_time" content="{{ .Format "2006-01-02T15:04:05Z0700" }}">{{ end }}
```
Comment (ryanfox1985) [inline layouts/partials/head.html]: format repeated many times, can you asign a variable and define just once?

Comment (salim-b) [reply layouts/partials/head.html]: you really want to introduce a variable for a function parameter??

File: exampleSite/config.toml (hunk 5 above)
Comment (salim-b) [reply exampleSite/config.toml]: ok

File: layouts/partials/head.html (hunk 1 above)
Comment (ryanfox1985) [reply layouts/partials/head.html]: It's not trolling, it's easy to review small and focus PR's

File: exampleSite/config.toml (hunk 5 above)
Comment (salim-b) [reply exampleSite/config.toml]: I've updated (and enhanced) the respective section of the README, see 6fe633b127c4fe9fd8c137161e3046a6208cae61 and 24c72b7108fa7159c02fc379162212e7edb063a4

File: layouts/partials/head.html (hunk 3 above)
Comment (salim-b) [reply layouts/partials/head.html]: I've added two variables to avoid this kind of redundancy, see dd81398e1fcc81655efe9600892804d1928ba1ba

File: layouts/partials/head.html (hunk 2 above)
Comment (salim-b) [reply layouts/partials/head.html]: > these meta changes are not related with the PR

Meanwhile it's actually an integral part of this PR since the same variable `$title_plain` is used across different meta tags.

**Comment by ryanfox1985** [comment]
Good refactors!
